### DISCUSSION TITLE
Use https://github.com/wire-elements/modal

### DISCUSSION THREAD

**Discussion by invaders-xx** [filamentphp/filament#1887] [Help] [answered]
Hi,
I am using https://github.com/wire-elements/modal in my existing project and I moved to your excellent package (many thanks) and I am wondering how to include the @livewire('livewire-ui-modal') in layout.

Many thanks for your help.

**Comment by cntabana** [comment]
The only way you can include the @livewire('livewire-ui-modal'), it's just to add it on this file https://github.com/laravel-filament/filament/blob/2.x/packages/admin/resources/views/components/layouts/base.blade.php#L31.

The problem is the file is in vendor which means each time you update the composer the @livewire('livewire-ui-modal') will be removed, so you will need to add it each time you upgrade the composer.

If you find another way just let me know.

**Reply by danharrin** [reply]
Haha please don't do this, for the reasons you have mentioned. Overriding vendor files is never the right solution.

**Reply by ziming** [reply]
Sometimes there is no choice (but not in this case). But if you do need to override vendor files. Use composer patches at least

https://github.com/cweagans/composer-patches

https://github.com/vaimo/composer-patches

**Reply by danharrin** [reply]
I promise you now, if you ever feel the need to override a Filament vendor file, LMK and I will tell you how to do the same thing without using a patch ;)

**Comment by danharrin** [comment]
You could create a `resources/views/vendor/filament/components/footer.blade.php` and put it there?

**Reply by cntabana** [reply]
Thank you @danharrin

**Comment by danharrin** [comment, marked as answer]
There's a better solution now: https://filamentphp.com/docs/2.x/admin/appearance#render-hooks

**Comment by R1NEJ0** [comment]
Hi! @invaders-xx .  Finally, did you make a modal using wire-element/modal? Could you paste a piece of your code, I will appreciate it.

Thanks a lot.`

**Reply by invaders-xx** [reply]
Hi, no I didn't yet, I am focusing on other stuff. But will let you know as soon as I get back to it.

**Reply by R1NEJ0** [reply]
Nice! 😃

**Reply by R1NEJ0** [reply]
@invaders-xx no need code example, I got it. 

Thanks anyway :)

**Reply by clerisonb** [reply]
@R1NEJ0 Do you mind sharing your solution?

Thanks.

**Reply by R1NEJ0** [reply]
@clerisonb about how to use wire-element?

**Reply by invaders-xx** [reply]
@R1NEJ0 Can you share how did you implement wire-element/modal? Could you paste a piece of your code, I will appreciate it. Many thanks in advance.

**Reply by R1NEJ0** [reply]
@invaders-xx yes of course, but I need a couple of days, I'm very busy these days. The main problem is wire-element CSS but I "fixed" it with a fudge.

**Comment by R1NEJ0** [comment]
Ok, I'm going to explain it as best as I can.

First, we need install dependecies of wire-element/modal.

`composer require wire-elements/modal`

And make vendor publish, this is for edit CSS of wire-element/modal.

`php artisan vendor:publish --tag=livewire-ui-modal-views`

After, add Render Hooks on `app\Providers\AppServiceProvider.php`, like [Filamentphp's documentation](https://filamentphp.com/docs/2.x/admin/appearance#render-hooks) says.

On top of class:

`use Filament\Facades\Filament;`
`use Illuminate\Support\Facades\Blade;`

And inside of function boot() of class AppServiceProvider:

`Filament::registerRenderHook(`
    ` 'body.start',`
    `  fn (): string => Blade::render('@livewire(\'livewire-ui-modal\')'),`
`);`

Ok, now we need create modal like [wire-elements/modal docs ](https://github.com/wire-elements/modal) says:

`php artisan make:livewire EditUser`

And inside of class created, replace `use Livewire\Component;` to `use LivewireUI\Modal\ModalComponent;` and extends to `ModalComponent` instead of `Component`. Here we can write the logic of our modal.

Perfect, now we need render buttons and actions, to do that I followed [filamentphp docs](https://filamentphp.com/docs/2.x/admin/pages#actions) and I added these lines on my page class:

On top of class I added this usage:
`use Filament\Pages\Actions\Action;`

And inside of class:

`protected function getActions(): array`
`{`
    `return [`
        `Action::make('settings')->action('openSettingsModal'),`
    `];`
`}`
 
`public function openSettingsModal(): void`
`{`
   ` $this->dispatchBrowserEvent('open-settings-modal');`
`}`

But to trigger modal you need replace `dispatchBrowserEvent` to `emit`, add `openModal` and replace `open-settings-modal` parameter to your modal name, in this case `edit-user`. It will be something like this:

`$this->emit('openModal', 'edit-user');`

And all is working, but be careful maybe You can't see it!! because it's going to render in bottom page so you need replace wire-element/modal CSS in `resources\views\vendor\livewire-ui-modal\modal.blade.php`

I attached my CSS configuration, **I've not made dark mode already.** So be careful and make a copy of original wire-element/modal vendor config/CSS. 

To use it you need replace extension from .txt to .blade.php .

[modal.txt](https://github.com/laravel-filament/filament/files/8665413/modal.txt)

I don't know if this is the correct way but it's working.

**Reply by R1NEJ0** [reply]
It would be cool if a tailwind expert wrote the code for the filament-style modals

**Reply by danharrin** [reply]
What inside of Filament causes the need to customise the view/css? Can we make changes to remove that customization step?

**Reply by R1NEJ0** [reply]
No idea, I think it's ok. There are not a lot of people making these custom modals so in my opinion it's not necessary. Your team are doing a big great work.  

When I work on a couple modals, I can post images.

**Reply by danharrin** [reply]
wire-elements/modal is a good package and it would be great to get it working out of the box :)

We do have our own modal utilities though... `<x-filament::modal>`

**Reply by R1NEJ0** [reply]
If I use <x-filament::modal> the modal doesn't render properly.

![image](https://user-images.githubusercontent.com/26463620/170891070-bfe26de6-d858-4d53-81b6-1d4e65cc2d3b.png)

**Reply by danharrin** [reply]
Does it have any content? It renders at the bottom of the page on mobile deliberately, I think.

**Reply by R1NEJ0** [reply]
Yes of course, It has content. In desktop it has the same behaivour.

**Reply by danharrin** [reply]
Do other modals in the app render ok?

**Reply by R1NEJ0** [reply]
Yes, all other modals are working perfectly.

**Reply by danharrin** [reply]
Whereabouts are you putting the code? Can you send the code? How are you opening the modal?

**Reply by R1NEJ0** [reply]
The code is exactly I put above, I'm not even a junior developer so maybe I'm not doing well.

**Reply by dandju** [reply]
Just a related question as I found this discussion after a quick search.

How would one pass the ID of the current row to the 'openSettingsModal' method from this example? 

So the ID can be passed on to the modal?

`$this->emit('openModal', 'edit-user', ['user_id' => $id, ...]);`

**Reply by R1NEJ0** [reply]
I left the project time ago, so I can't help you. I didn't remember.

But I found this discussion, maybe you can try -> https://github.com/filamentphp/filament/discussions/3419

And it's on filament tricks website -> https://filamentphp.com/tricks/opening-a-modal-in-a-blade-with-button-outside-its-trigger-slot

With this method maybe you can pass all data as you want.

Sorry about late reply.
